### PR TITLE
ARROW-8619: [C++] Use distinct enum values for MonthInterval, DayTimeInterval

### PR DESCRIPTION
**Pull request by wesm**
This enables us to eliminate "special" handling of `Type::INTERVAL` since these types have a different internal data representation. The deleted code in this PR is evidence of this. 

This is a breaking API change but since this is a lesser-used part of the project I don't think that many users will be impacted.

### CODE REVIEW

**Comment by wesm** [comment]
Will fix the builds

File: r/R/enums.R [hunk 1]
```diff
@@ -42,12 +42,35 @@ DateUnit <- enum("DateUnit", DAY = 0L, MILLI = 1L)
...
+  DOUBLE = 12L,
+  STRING = 13L,
+  BINARY = 14L,
+  FIXED_SIZE_BINARY = 15L,
+  DATE32 = 16L,
+  DATE64 = 17L,
+  TIMESTAMP = 18L,
+  TIME32 = 19L,
+  TIME64 = 20L,
+  INTERVAL_MONTHS = 21L,
```
Comment (emkornfield) [inline r/R/enums.R:66]: are these enums stored anyplace?

**Comment by emkornfield** [comment]
In principal this looks ok to me.  It seems like there is at least one place that was missed /arrow/cpp/src/gandiva/jni/expression_registry_helper.cc:141:10: error: ‘INTERVAL’ is not a member of ‘arrow::Type::type’

As long as the flatbuffers doesn't change, I'm OK with this.  Might be worth sending a mail out to dev@ just to make people aware of it?

**Comment by wesm** [comment]
For some reason I can't get JNI running in my local setup

```
CMake Error at /home/wesm/cpp-toolchain/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find JNI (missing: JAVA_AWT_INCLUDE_PATH)
Call Stack (most recent call first):
  /home/wesm/cpp-toolchain/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /home/wesm/cpp-toolchain/share/cmake-3.16/Modules/FindJNI.cmake:372 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  src/gandiva/jni/CMakeLists.txt:27 (find_package)
```

tried with 

```
export JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64
```

and 

```
export JAVA_HOME=/usr/lib/jvm/java-11-openjdk-amd64
```

It worked with

```
export JAVA_HOME=/usr/lib/jvm/java-14-oracle
```

though

**Comment by wesm** [comment]
The ursabot build failures are spurious

**Comment by kszucs** [comment]
@ursabot build

**Comment by kszucs** [comment]
> The ursabot build failures are spurious

Occasionally happens after a force push.

**Review by fsaintjacques — APPROVED** [review]

**Comment by wesm** [comment]
+1

**Pull request closed without merging**